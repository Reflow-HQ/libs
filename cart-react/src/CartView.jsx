import React, { useState, useEffect } from "react";

import { ShoppingCartProvider, useShoppingCart } from "./CartContext";
import CartSlide from "./steps/CartSlide";
import CheckoutSlide from "./steps/CheckoutSlide";

const CartView = ({ cart, auth, successURL, cancelURL, onMessage }) => {
  return (
    <ShoppingCartProvider cart={cart} auth={auth}>
      <CartUI successURL={successURL} cancelURL={cancelURL} onMessage={onMessage} />
    </ShoppingCartProvider>
  );
};

export function CartUI({ successURL, cancelURL, onMessage }) {
  const [step, setStep] = useState("cart");

  const cart = useShoppingCart();

  const { products, isLoading, isLoaded, isUnavailable, t } = cart;

  const errorMessage = isUnavailable
    ? t("cart.errors.unavailable")
    : isLoaded
    ? !cart.arePaymentProvidersAvailable()
      ? t("cart.errors.no_payment_methods")
      : cart.onlyPaypalNoDelivery()
      ? t("cart.errors.only_paypal_no_delivery")
      : !products.length
      ? t("cart.errors.empty")
      : ""
    : "";

  console.log("render CartUI", isLoaded);
  console.log("render CartUI", errorMessage);

  useEffect(() => {
    if (!products.length && step !== "cart") {
      setStep("cart");
    }
  }, [products]);

  function renderActiveSlide() {
    if (step === "cart") {
      return (
        <CartSlide
          step={step}
          setStep={setStep}
          successURL={successURL}
          cancelURL={cancelURL}
          onMessage={onMessage}
        />
      );
    } else {
      return (
        <CheckoutSlide
          step={step}
          setStep={setStep}
          successURL={successURL}
          cancelURL={cancelURL}
          onMessage={onMessage}
        />
      );
    }
  }

  /* Todo: write tests with jest. I need to have a variant that uses a static state as well. I also need
    to show different attributes for customizing panels, layout, injecting components, placing things in the URL(?)

   */

  return (
    <div className="reflow-shopping-cart">
      <div className={"ref-loading-backdrop" + (isLoading ? " active" : "")}></div>
      {errorMessage ? (
        <div className="ref-message">{errorMessage}</div>
      ) : isLoaded ? (
        renderActiveSlide()
      ) : null}
    </div>
  );
}

export default CartView;
