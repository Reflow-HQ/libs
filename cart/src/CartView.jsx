import { useState } from "react";

import { useShoppingCart } from "./CartContext";
import CartSlide from "./steps/CartSlide";
import CheckoutSlide from "./steps/CheckoutSlide";

const CartView = () => {
  const [step, setStep] = useState("cart");

  const { cartState, t } = useShoppingCart();
  const { products, isLoading } = cartState;

  function renderActiveSlide() {
    if (step === "cart") {
      return <CartSlide setStep={setStep} />;
    } else {
      return <CheckoutSlide setStep={setStep} />;
    }
  }

  /* Todo: write tests with jest. I need to have a variant that uses a static state as well. I also need
    to show different attributes for customizing panels, layout, injecting components, placing things in the URL(?)

   */

  return (
    <div className="reflow-shopping-cart">
      <div className={"ref-loading-backdrop " + (isLoading ? "active" : "")}></div>
      {products.length ? (
        renderActiveSlide()
      ) : (
        <div className="ref-message">{t("cart.errors.empty")}</div>
      )}
    </div>
  );
};

export default CartView;
