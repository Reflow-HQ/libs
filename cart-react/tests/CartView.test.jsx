/**
 * @jest-environment jsdom
 */

import React from "react";

// import renderer from "react-test-renderer";
import { render, renderHook, screen, act, waitFor } from "@testing-library/react";

import CartView, { useCart } from "../";

let products = [
  {
    id: "123",
    lineItemID: "456",
    type: "physical",
    inStock: true,
    quantity: 1,
  },
];

const storeID = "1234";
const cartKey = "key";
const config = {
  storeID,
  apiBase: "http://api.reflow.local/v1",
};

const defaultCartContent = {
  isLoading: false,
  isLoaded: true,
  isUnavailable: false,
};

function mockFetch() {
  return (url) => {
    let response;

    switch (url) {
      case `http://api.reflow.local/v1/stores/${storeID}/carts/`: {
        response = { cartKey };
        break;
      }
      case `http://api.reflow.local/v1/stores/${storeID}/carts/${cartKey}`: {
        response = { ...defaultCartContent, products };
        break;
      }
      default: {
        return Promise.resolve({
          status: 404,
          json: () => Promise.resolve({}),
        });
      }
    }

    return Promise.resolve({
      status: 200,
      ok: true,
      json: () => Promise.resolve(response),
    });
  };
}
beforeAll(() => {
  // we're using fake timers because we don't want to
  // wait a full second for this test to run.
  jest.useFakeTimers();
});

afterAll(() => {
  jest.useRealTimers();
});

beforeEach(() => {
  global.fetch = jest.fn(mockFetch());
});

afterEach(() => {
  global.fetch.mockClear();
});

it("renders a loading screen", async () => {
  // let cart;

  // act(() => {
  // renderHook(() => {
  // });
  // cart = useCart(config);

  const { result } = renderHook(() => useCart(config));

  render(<CartView cart={result.current} />);

  jest.advanceTimersByTime(1000);
  // await waitFor(() => expect(fetch).toHaveBeenCalledTimes(1));

  expect(await screen.findByTestId("cart")).toBeInTheDocument();
  // const component = renderer.create(<CartView cart={cart} />);
  // expect(component.toJSON()).toMatchSnapshot();

  // Todo: implement interactive snapshot testing
  // https://jestjs.io/docs/snapshot-testing
});
