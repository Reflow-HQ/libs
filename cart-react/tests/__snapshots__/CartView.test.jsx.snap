// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a loading screen 1`] = `
<div
  className="reflow-shopping-cart"
>
  <div
    className="ref-loading-backdrop"
  />
</div>
`;

exports[`renders the cart slide 1`] = `
<div
  className="reflow-shopping-cart"
>
  <div
    className="ref-loading-backdrop"
  />
</div>
`;
